lib/tests/modules/types.nix

```
@@ -21,5 +26,328 @@ in
+        (oneOf [
+          str
+          int
+        ]).description == "string or signed integer";
+      assert
+        (oneOf [
+          str
+          int
+          bool
+        ]).description == "string or signed integer or boolean";
+      assert
+        (oneOf [
+          (listOf str)
+          int
+          bool
+        ]).description == "(list of string) or signed integer or boolean";
+      assert
+        (oneOf [
+          ints.positive
+          str
+        ]).description == "positive integer, meaning >0, or string";
+      assert (passwdEntry str).description == "string, not containing newlines or colons";
+      assert (passwdEntry int).description == "signed integer, not containing newlines or colons";
+      assert (passwdEntry bool).description == "boolean, not containing newlines or colons";
+      assert
+        (passwdEntry (either int str)).description
+        == "(signed integer or string), not containing newlines or colons";
+      assert
+        (passwdEntry (nullOr str)).description == "(null or string), not containing newlines or colons";
+      assert
+        (passwdEntry (listOf str)).description == "(list of string), not containing newlines or colons";
+      assert
+        (passwdEntry (attrsOf int)).description
+        == "(attribute set of signed integer), not containing newlines or colons";
+      assert
+        (passwdEntry ints.positive).description
+        == "(positive integer, meaning >0), not containing newlines or colons";
+      assert (pathWith { }).description == "path";
+      assert (pathWith { absolute = true; }).description == "absolute path";
+      assert (pathWith { inStore = true; }).description == "path in the Nix store";
+      assert
+        (pathWith {
+          absolute = true;
+          inStore = true;
+        }).description == "absolute path in the Nix store";
+      assert (pathWith { absolute = false; }).description == "relative path";
+      assert (pathWith { absolute = null; }).description == "path";
+      assert (pathWith { inStore = false; }).description == "path not in the Nix store";
+      assert (pathWith { inStore = null; }).description == "path";
+      assert (separatedString "").description == "Concatenated string";
+      assert (separatedString ",").description == "strings concatenated with \",\"";
+      assert (separatedString "\n").description == ''strings concatenated with "\n"'';
+      assert (separatedString ":").description == "strings concatenated with \":\"";
+      assert (strMatching "[a-z]+").description == "string matching the pattern [a-z]+";
+      assert
+        (strMatching "[0-9]{3}-[0-9]{2}-[0-9]{4}").description
+        == "string matching the pattern [0-9]{3}-[0-9]{2}-[0-9]{4}";
+      assert (strMatching ".*\\.txt").description == "string matching the pattern .*\\.txt";
+      assert
+        (submodule { freeformType = attrsOf int; }).description
+        == "open submodule of attribute set of signed integer";
```

This seems to describe a wildcard-like feature, an individual item type, but it's more of a union type.
```suggestion
        == "open submodule over attribute set of signed integer";
```

```suggestion
        == "union of submodule and attribute set of signed integer";
```

---

I don't like the phrase union. I like to  think of it as an intersection (`&`), not a union (`|`).

It must satisfy both side `submodule with options A,B AND (&) optional additional fields of type C`

---

The freeformType extends the module somewhat like a union type (not a sum!), but the module can alter the freeformType arbitrarily, intersection, union, or unrelated, e.g. attrsOf str, with an int option.
I only had the prior in mind, but clearly it's not such a great model, because unions are supposed to be symmetrical.
So let's not use these set terms :+1: 

`extending` is also not accurate, because the submodule may alter types, making them unrelated in terms of subtyping, and `extending` suggests subtyping (or Liskov principle in OO terms).

So then I think `open submodule over` remains.
Haven't thought of something nicer. Only `submodule augmenting` came close, but it's not as good.

---

While I appreciate the effort to use accurate terminology, I'd caution against getting overly technical in option-type descriptions, as these are _very_ user-facing and are consumed by users with _all_ levels of experience.

I think the original "open submodule **of** attribute set of signed integer" is fine, but I appreciate if you want to avoid the inaccuracy.

Some alternative ideas to throw at the proverbial wall:

```suggestion
        == "submodule or attribute set of signed integer";
```
(could be confused with an `either` union)

```suggestion
        == "open submodule permitting attribute set of signed integer";
```


```suggestion
        == "freeform submodule of attribute set of signed integer";
```

```suggestion
        == "freeform submodule allowing attribute set of signed integer";
```

```suggestion
        == "freeform submodule using attribute set of signed integer";
```

```suggestion
        == "open submodule using attribute set of signed integer";
```

---

Okay i would like to conclude this. I think its probably fine to go with the current `open submodule of attribute set of signed integer` If we figure that people tend to misunderstand this, we can still improve it.

The technical "correct" description is probably to verbose and hard to understand.
While the other suggestions are hard to agree on.